File: blockchain/src/main/java/pt/um/lei/masb/blockchain/data/HumidityData.java
package pt.um.lei.masb.blockchain.data;

import pt.um.lei.masb.blockchain.Sizeable;

import javax.persistence.Basic;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.util.Objects;


/**
 * Humidity data can be expressed in Absolute/Volumetric or Relative humidity.
 * As such possible measurements can be in g/kg, Kg/kg or percentage.
 */
@Entity
public final class HumidityData extends GeoData implements Sizeable {
    @Id
    @GeneratedValue
    private long id;

    @Basic(optional = false)
    private double hum;

    @Basic(optional = false)
    private HUnit unit;

    public HumidityData(double hum,
                        @NotNull HUnit unit,
                        BigDecimal lat,
                        BigDecimal lng) {
        super(lat, lng);
        this.hum = hum;
        this.unit = unit;
    }

    protected HumidityData() {
        super(new BigDecimal(0), new BigDecimal(0));
    }

    public double convertToGbyKG() {
        var res = hum;
        switch(unit) {
            case G_BY_KG:
                break;
            case KG_BY_KG:
                res *= 1000;
                break;
            case RELATIVE:
                break;
        }
        return res;
    }

    public double convertToKGbyKG() {
        var res = hum;
        switch(unit) {
            case G_BY_KG:
                res /= 1000;
                break;
            case KG_BY_KG:
                break;
            case RELATIVE:
                break;
        }
        return res;
    }

    public double getHum() {
        return hum;
    }

    public HUnit getUnit() {
        return unit;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        HumidityData that = (HumidityData) o;
        return Double.compare(that.hum, hum) == 0 &&
                unit == that.unit;
    }

    @Override
    public int hashCode() {

        return Objects.hash(hum, unit);
    }

    @Override
    public @NotNull String toString() {
        return "HumidityData{" +
                "hum=" + hum +
                ", unit=" + unit +
                '}';
    }
}

File: blockchain/src/main/kotlin/pt/um/lei/masb/blockchain/data/HumidityData.kt
package pt.um.lei.masb.blockchain.data

import com.orientechnologies.orient.core.record.OElement
import pt.um.lei.masb.blockchain.Coinbase
import pt.um.lei.masb.blockchain.Hash
import pt.um.lei.masb.blockchain.persistance.NewInstanceSession
import pt.um.lei.masb.blockchain.utils.Crypter
import java.io.InvalidClassException
import java.math.BigDecimal


/**
 * Humidity data can be expressed in Absolute/Volumetric
 * or Relative humidity. As such possible measurements
 * can be in g/Kg ([HUnit.G_BY_KG]), Kg/Kg ([HUnit.KG_BY_KG])
 * or percentage ([HUnit.RELATIVE]) expressed by the [unit].
 */
data class HumidityData(
    val hum: BigDecimal,
    val unit: HUnit
) : BlockChainData {
    override fun digest(c: Crypter): Hash =
        c.applyHash(
            """
            $hum
            ${unit.name}
            ${unit.ordinal}
            """.trimIndent()
        )


    override fun store(
        session: NewInstanceSession
    ): OElement =
        session
            .newInstance("Humidity")
            .let {
                it.setProperty("hum", hum)
                val hUnit = when (unit) {
                    HUnit.G_BY_KG -> 0x00.toByte()
                    HUnit.KG_BY_KG -> 0x01.toByte()
                    HUnit.RELATIVE -> 0x02.toByte()
                }
                it.setProperty("unit", hUnit)
                it
            }


    override fun calculateDiff(previous: SelfInterval): BigDecimal =
        when (previous) {
            is HumidityData ->
                calculateDiffHum(previous)
            else ->
                throw InvalidClassException("SelfInterval supplied is not ${this::class.simpleName}")
        }

    private fun calculateDiffHum(previous: HumidityData): BigDecimal {
        val newH: BigDecimal
        val oldH: BigDecimal
        if (unit == HUnit.RELATIVE) {
            newH = hum
            oldH = previous.hum
        } else {
            newH = unit.convertTo(hum, HUnit.KG_BY_KG)
            oldH = previous.unit.convertTo(hum, HUnit.KG_BY_KG)
        }
        return newH.subtract(oldH)
            .divide(oldH, Coinbase.MATH_CONTEXT)
    }


    override fun toString(): String {
        return "HumidityData(hum = $hum, unit = $unit)"
    }

}